lib/src/codepage.dart

```diff
@@ -0,0 +1,486 @@
+  /// If not true, the bytes must be calid and defined characters.
+  String convert(List<int> input, {bool allowInvalid = false});
+
+  _CodePageEncoder _createEncoder();
+  int _char(int byte);
+}
+
+/// Creates a decoder from [characters].
+///
+/// Recognizes if [characters] contains only characters in the BMP,
+/// and creates a [BmpCodePageDecoder] in that case.
+CodePageDecoder _createDecoder(String characters) {
+  var result = Uint32List(256);
+  var i = 0;
+  var allChars = 0;
+  for (var char in characters.runes) {
+    if (i >= 256) {
+      throw ArgumentError.value(
+          characters, "characters", "Must contain 256 characters");
+    }
+    result[i] = char;
+    allChars |= char;
+  }
+  if (i < 256) {
+    throw ArgumentError.value(
+        characters, "characters", "Must contain 256 characters");
+  }
+  if (allChars <= 0xFFFF) {
+    // It's in the BMP.
+    return _BmpCodePageDecoder(characters);
+  }
+  return _NonBmpCodePageDecoder._(result);
+}
+
+/// Code page with non-BMP characters.
+class _NonBmpCodePageDecoder extends Converter<List<int>, String>
+    implements CodePageDecoder {
+  final Uint32List _characters;
+  _NonBmpCodePageDecoder(String characters)
+      : _characters = _buildMapping(characters);
+  _NonBmpCodePageDecoder._(this._characters);
+
+  int _char(int byte) => _characters[byte];
+
+  static Uint32List _buildMapping(String characters) {
+    var result = Uint32List(256);
+    var i = 0;
+    for (var char in characters.runes) {
+      if (i >= 256) {
+        throw ArgumentError.value(
+            characters, "characters", "Must contain 256 characters");
+      }
+      result[i++] = char;
+    }
+    if (i < 256) {
+      throw ArgumentError.value(
+          characters, "characters", "Must contain 256 characters");
+    }
+    return result;
+  }
+
+  _CodePageEncoder _createEncoder() {
+    var result = <int, int>{};
+    for (var i = 0; i < 256; i++) {
+      var char = _characters[i];
+      if (char != 0xFFFD) {
+        result[char] = i;
+      }
+    }
+    return _CodePageEncoder._(result);
+  }
+
+  String convert(List<int> input, {bool allowInvalid = false}) {
+    var buffer = Uint32List(input.length);
+    for (var i = 0; i < input.length; i++) {
+      var byte = input[i];
+      if (byte & 0xff != byte) throw FormatException("Not a byte", input, i);
+      buffer[i] = _characters[byte];
+    }
+    return String.fromCharCodes(buffer);
+  }
+}
+
+class _BmpCodePageDecoder extends Converter<List<int>, String>
+    implements CodePageDecoder {
+  final String _characters;
+  _BmpCodePageDecoder(String characters) : _characters = characters {
+    if (characters.length != 256) {
+      throw ArgumentError.value(
+          characters, "characters", "Must contain 256 characters. Was ${characters.length}");
+    }
+  }
+
+  int _char(int byte) => _characters.codeUnitAt(byte);
+
+  /// Decodes bytes into characters using this code page.
```

Comment:
Remove copy/paste comment?

We should use `@override` in this package to make this more clear.

Reply:
Comment removed.